File: lib/iolib/iolib.go
package iolib

import (
	"bufio"
	"errors"
	"fmt"
	"io"
	"os"
	"os/exec"
	"runtime"
	"syscall"

	"github.com/arnodel/golua/lib/packagelib"
	rt "github.com/arnodel/golua/runtime"
)

// BufferedStdFiles sets wether std files should be buffered
var BufferedStdFiles bool = true

type ioKeyType struct{}

var ioKey = rt.AsValue(ioKeyType{})

// LibLoader can load the io lib.
var LibLoader = packagelib.Loader{
	Load: load,
	Name: "io",
}

func load(r *rt.Runtime) (rt.Value, func()) {
	methods := rt.NewTable()

	meta := rt.NewTable()
	r.SetEnv(meta, "__name", rt.StringValue("file"))
	r.SetEnv(meta, "__index", rt.TableValue(methods))

	rt.SolemnlyDeclareCompliance(
		rt.ComplyCpuSafe|rt.ComplyMemSafe|rt.ComplyIoSafe,

		r.SetEnvGoFunc(methods, "read", fileread, 1, true),
		r.SetEnvGoFunc(methods, "lines", filelines, 1, true),
		r.SetEnvGoFunc(methods, "close", fileclose, 1, false),
		r.SetEnvGoFunc(methods, "flush", fileflush, 1, false),
		r.SetEnvGoFunc(methods, "seek", fileseek, 3, false),
		r.SetEnvGoFunc(methods, "setvbuf", filesetvbuf, 3, false),
		// TODO: setvbuf,
		r.SetEnvGoFunc(methods, "write", filewrite, 1, true),

		r.SetEnvGoFunc(meta, "__close", file__close, 1, false),
	)

	rt.SolemnlyDeclareCompliance(
		rt.ComplyCpuSafe|rt.ComplyMemSafe|rt.ComplyTimeSafe|rt.ComplyTimeSafe|rt.ComplyIoSafe,

		r.SetEnvGoFunc(meta, "__tostring", tostring, 1, false),
	)

	var (
		stdoutOpts = statusNotClosable
		stderrOpts = statusNotClosable
		stdinOpts  = statusNotClosable
	)
	if BufferedStdFiles {
		stdoutOpts |= bufferedWrite
		stdinOpts |= bufferedRead
	}

	stdinFile := NewFile(os.Stdin, stdinOpts)
	stdoutFile := NewFile(os.Stdout, stdoutOpts)
	stderrFile := NewFile(os.Stderr, stderrOpts)
	// This is not a good pattern - it has to do for now.
	if r.Stdout == nil {
		r.Stdout = stdoutFile.writer
	}
	stdin := newFileUserData(stdinFile, meta)
	stdout := newFileUserData(stdoutFile, meta)
	stderr := newFileUserData(stderrFile, meta) // I''m guessing, don't buffer stderr?

	r.SetRegistry(ioKey, rt.AsValue(&ioData{
		defaultOutput: stdout,
		defaultInput:  stdin,
		metatable:     meta,
	}))
	pkg := rt.NewTable()
	r.SetEnv(pkg, "stdin", rt.UserDataValue(stdin))
	r.SetEnv(pkg, "stdout", rt.UserDataValue(stdout))
	r.SetEnv(pkg, "stderr", rt.UserDataValue(stderr))

	rt.SolemnlyDeclareCompliance(
		rt.ComplyCpuSafe|rt.ComplyMemSafe|rt.ComplyIoSafe,

		r.SetEnvGoFunc(pkg, "close", ioclose, 1, false),
		r.SetEnvGoFunc(pkg, "flush", ioflush, 0, false),
		r.SetEnvGoFunc(pkg, "input", input, 1, false),
		r.SetEnvGoFunc(pkg, "lines", iolines, 1, true),
		r.SetEnvGoFunc(pkg, "open", open, 2, false),
		r.SetEnvGoFunc(pkg, "output", output, 1, false),
		r.SetEnvGoFunc(pkg, "popen", popen, 2, false),
		r.SetEnvGoFunc(pkg, "read", ioread, 0, true),
		r.SetEnvGoFunc(pkg, "tmpfile", tmpfile, 0, false),
		r.SetEnvGoFunc(pkg, "write", iowrite, 0, true),
	)

	rt.SolemnlyDeclareCompliance(
		rt.ComplyCpuSafe|rt.ComplyMemSafe|rt.ComplyTimeSafe|rt.ComplyIoSafe,

		r.SetEnvGoFunc(pkg, "type", typef, 1, false),
	)

	// This function should make sure known buffers are flushed before quitting
	var cleanup = func() {
		getIoData(r).defaultOutputFile().Flush()
		stdoutFile.Flush()
		stderrFile.Flush()
	}

	return rt.TableValue(pkg), cleanup
}

type ioData struct {
	defaultOutput *rt.UserData
	defaultInput  *rt.UserData
	metatable     *rt.Table
}

func getIoData(r *rt.Runtime) *ioData {
	return r.Registry(ioKey).Interface().(*ioData)
}

func (d *ioData) defaultOutputFile() *File {
	return d.defaultOutput.Value().(*File)
}

func (d *ioData) defaultInputFile() *File {
	return d.defaultInput.Value().(*File)
}

func pushingNextIoResult(r *rt.Runtime, c *rt.GoCont, ioErr error) (rt.Cont, error) {
	next := c.Next()
	if ioErr != nil {
		return r.ProcessIoError(next, ioErr)
	}
	r.Push1(next, rt.BoolValue(true))
	return next, nil
}

func ioclose(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	var f *File
	if c.NArgs() == 0 {
		f = getIoData(t.Runtime).defaultOutputFile()
	} else {
		var err error
		f, err = FileArg(c, 0)
		if err != nil {
			return nil, err
		}
	}

	var cont rt.Cont
	var err error
	if f.file == nil {
		cont, err = f.close(t, c)
	} else {
		cont, err = pushingNextIoResult(t.Runtime, c, f.Close())
	}

	return cont, err
}

func fileclose(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	return ioclose(t, c)
}

func file__close(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	f, err := FileArg(c, 0)
	if err != nil {
		return nil, err
	}
	closeErr := f.Close()
	_ = closeErr // TODO: something with the error
	return c.Next(), nil
}

func ioflush(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	var f *File
	if c.NArgs() == 0 {
		f = getIoData(t.Runtime).defaultOutputFile()
	} else {
		var err error
		f, err = FileArg(c, 0)
		if err != nil {
			return nil, err
		}
	}
	return pushingNextIoResult(t.Runtime, c, f.Flush())
}

func fileflush(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	return ioflush(t, c)
}

func errFileOrFilename() error {
	return errors.New("#1 must be a file or a filename")
}

func input(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	ioData := getIoData(t.Runtime)
	if c.NArgs() == 0 {
		return c.PushingNext1(t.Runtime, rt.UserDataValue(ioData.defaultInput)), nil
	}
	var (
		fv  *rt.UserData
		arg = c.Arg(0)
	)
	switch arg.Type() {
	case rt.StringType:
		f, ioErr := OpenFile(t.Runtime, arg.AsString(), "r")
		if ioErr != nil {
			return nil, ioErr
		}
		fv = newFileUserData(f, ioData.metatable)
	case rt.UserDataType:
		_, err := FileArg(c, 0)
		if err != nil {
			return nil, errFileOrFilename()
		}
		fv = arg.AsUserData()
	default:
		return nil, errFileOrFilename()
	}
	ioData.defaultInput = fv
	return c.PushingNext1(t.Runtime, rt.UserDataValue(fv)), nil
}

func output(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	ioData := getIoData(t.Runtime)
	if c.NArgs() == 0 {
		return c.PushingNext1(t.Runtime, rt.UserDataValue(ioData.defaultOutput)), nil
	}
	var (
		fv  *rt.UserData
		arg = c.Arg(0)
	)
	switch arg.Type() {
	case rt.StringType:
		f, ioErr := OpenFile(t.Runtime, arg.AsString(), "w")
		if ioErr != nil {
			return nil, ioErr
		}
		fv = newFileUserData(f, ioData.metatable)
	case rt.UserDataType:
		_, err := FileArg(c, 0)
		if err != nil {
			return nil, errFileOrFilename()
		}
		fv = arg.AsUserData()
	default:
		return nil, errFileOrFilename()
	}
	// Make sure the current output is flushed
	ioData.defaultOutput.Value().(*File).Flush()
	ioData.defaultOutput = fv
	return c.PushingNext1(t.Runtime, rt.UserDataValue(fv)), nil
}

func iolines(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	var (
		f         *File
		eofAction = closeAtEOF
	)
	if c.NArgs() == 0 || c.Arg(0) == rt.NilValue {
		f = getIoData(t.Runtime).defaultInputFile()
		eofAction = doNotCloseAtEOF
	} else {
		fname, err := c.StringArg(0)
		if err != nil {
			return nil, err
		}
		var ioErr error
		f, ioErr = OpenFile(t.Runtime, string(fname), "r")
		if ioErr != nil {
			return nil, ioErr
		}
	}
	readers, fmtErr := getFormatReaders(c.Etc())
	if fmtErr != nil {
		return nil, fmtErr
	}
	return c.PushingNext(t.Runtime, rt.FunctionValue(lines(t.Runtime, f, readers, eofAction))), nil
}

func filelines(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	f, err := FileArg(c, 0)
	if err != nil {
		return nil, err
	}
	readers, fmtErr := getFormatReaders(c.Etc())
	if fmtErr != nil {
		return nil, fmtErr
	}

	return c.PushingNext(t.Runtime, rt.FunctionValue(lines(t.Runtime, f, readers, doNotCloseAtEOF))), nil
}

const (
	closeAtEOF      = 1
	doNotCloseAtEOF = 0
)

func lines(r *rt.Runtime, f *File, readers []formatReader, flags int) *rt.GoFunction {
	if len(readers) == 0 {
		readers = []formatReader{lineReader(false)}
	}
	iterator := func(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
		next := c.Next()
		// if f.closed {
		// 	return next, nil
		// }
		err := read(r, f, readers, next)
		if err != nil {
			if err == io.EOF {
				if flags&closeAtEOF != 0 {
					if err := f.Close(); err != nil {
						return t.ProcessIoError(next, err)
					}
				}
				t.Push1(next, rt.NilValue)
				return next, nil
			}
			return nil, err
		}
		return next, nil
	}
	iterGof := rt.NewGoFunction(iterator, "linesiterator", 0, false)
	iterGof.SolemnlyDeclareCompliance(rt.ComplyCpuSafe | rt.ComplyMemSafe | rt.ComplyIoSafe)
	return iterGof

}

func open(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	fname, err := c.StringArg(0)
	if err != nil {
		return nil, err
	}
	mode := "r"
	if c.NArgs() >= 2 {
		mode, err = c.StringArg(1)
		if err != nil {
			return nil, err
		}
	}
	f, ioErr := OpenFile(t.Runtime, fname, mode)
	if ioErr != nil {
		return pushingNextIoResult(t.Runtime, c, ioErr)
	}

	u := newFileUserData(f, getIoData(t.Runtime).metatable)
	return c.PushingNext(t.Runtime, rt.UserDataValue(u)), nil
}

func popen(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	cmdStr, err := c.StringArg(0)
	if err != nil {
		return nil, err
	}

	mode := "r"
	if c.NArgs() >= 2 {
		mode, err = c.StringArg(1)
		if err != nil {
			return nil, err
		}
	}

	var cmdArgs []string
	if runtime.GOOS == "windows" {
		cmdArgs = []string{"C:\\Windows\\system32\\cmd.exe", "/c", cmdStr}
	} else {
		cmdArgs = []string{"/bin/sh", "-c", cmdStr}
	}

	cmd := exec.Cmd{
		Path: cmdArgs[0],
		Args: cmdArgs,
	}

	outDummy, inDummy, err := os.Pipe()
	if err != nil {
		return nil, err
	}

	f := &File{
		writer: &nobufWriter{inDummy},
		reader: &nobufReader{outDummy},
		name: cmdStr,
	}

	var stdout io.ReadCloser
	var stdin io.WriteCloser
	switch mode {
		case "r":
			stdout, err = cmd.StdoutPipe()
			f.reader = bufio.NewReader(stdout)
		case "w":
			stdin, err = cmd.StdinPipe()
			f.writer = bufio.NewWriterSize(stdin, 65536)
		default:
			err = errors.New("invalid mode")
	}
	// called *only* from io.close
	f.close = func(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
		if stdout != nil {
			err = stdout.Close()
			if err != nil {
				return pushingNextIoResult(t.Runtime, c, err)
			}
		}
		if stdin != nil {
			err := stdin.Close()
			if err != nil {
				return pushingNextIoResult(t.Runtime, c, err)
			}
		}

		cont := c.Next()

		cmd.Wait()
		ps := cmd.ProcessState
		if ps.Success() {
			t.Runtime.Push(cont, rt.BoolValue(true))
		} else {
			t.Runtime.Push(cont, rt.NilValue)
		}

		exit := rt.StringValue("exit")
		code := rt.IntValue(int64(ps.ExitCode()))
		if !ps.Exited() {
			// received signal instead of normal exit
			exit = rt.StringValue("signal")
			if runtime.GOOS != "windows" {
				// i can't find out what Sys() is on windows ...
				ws := ps.Sys().(syscall.WaitStatus)
				sig := ws.Signal()
				code = rt.IntValue(int64(sig)) // syscall.Signal
			}
		}

		t.Runtime.Push(cont, exit, code)

		return c.Next(), nil
	}

	if err != nil {
		return pushingNextIoResult(t.Runtime, c, err)
	}

	err = cmd.Start()
	if err != nil {
		return nil, err
	}

	u := newFileUserData(f, getIoData(t.Runtime).metatable)
	return c.PushingNext(t.Runtime, rt.UserDataValue(u)), nil
}

func typef(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	f, err := FileArg(c, 0)
	var val rt.Value
	if err != nil {
		val = rt.NilValue
	} else if f.IsClosed() {
		val = rt.StringValue("closed file")
	} else {
		val = rt.StringValue("file")
	}
	return c.PushingNext(t.Runtime, val), nil
}

func iowrite(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	return write(t.Runtime, rt.UserDataValue(getIoData(t.Runtime).defaultOutput), c)
}

func filewrite(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	return write(t.Runtime, c.Arg(0), c)
}

func write(r *rt.Runtime, vf rt.Value, c *rt.GoCont) (rt.Cont, error) {
	f, ok := ValueToFile(vf)
	if !ok {
		return nil, errors.New("#1 must be a file")
	}
	if f.IsClosed() {
		return nil, errFileAlreadyClosed
	}
	var err error
	for _, val := range c.Etc() {
		switch val.Type() {
		case rt.StringType:
		case rt.IntType:
		case rt.FloatType:
		default:
			return nil, errors.New("argument must be a string or a number")
		}
		s, _ := val.ToString()
		if err = f.WriteString(s); err != nil {
			break
		}
	}
	next := c.Next()
	if err != nil {
		return r.ProcessIoError(next, err)
	} else {
		r.Push(next, vf)
	}
	return next, nil
}

func fileseek(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	f, err := FileArg(c, 0)
	if err != nil {
		return nil, err
	}
	whence := io.SeekCurrent
	offset := int64(0)
	nargs := c.NArgs()
	if nargs >= 2 {
		whenceName, err := c.StringArg(1)
		if err != nil {
			return nil, err
		}
		switch whenceName {
		case "cur":
			whence = io.SeekCurrent
		case "set":
			whence = io.SeekStart
		case "end":
			whence = io.SeekEnd
		default:
			return nil, errors.New(`#1 must be "cur", "set" or "end"`)
		}
	}
	if nargs >= 3 {
		offsetI, err := c.IntArg(2)
		if err != nil {
			return nil, err
		}
		offset = int64(offsetI)
	}
	pos, ioErr := f.Seek(offset, whence)
	next := c.Next()
	if ioErr != nil {
		t.Push1(next, rt.NilValue)
		t.Push1(next, rt.StringValue(ioErr.Error()))
	} else {
		t.Push1(next, rt.IntValue(pos))
	}
	return next, nil
}

func filesetvbuf(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	if err := c.CheckNArgs(2); err != nil {
		return nil, err
	}
	f, err := FileArg(c, 0)
	if err != nil {
		return nil, err
	}
	mode, err := c.StringArg(1)
	if err != nil {
		return nil, err
	}
	var size int64
	if c.NArgs() > 2 {
		size, err = c.IntArg(2)
		if err != nil {
			return nil, err
		}
	}
	bufErr := f.SetWriteBuffer(mode, int(size))
	if bufErr != nil {
		return nil, bufErr
	}
	return c.PushingNext1(t.Runtime, rt.BoolValue(true)), nil
}

func tmpfile(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	f, err := TempFile(t.Runtime)
	if err != nil {
		return nil, err
	}
	fv := newFileUserData(f, getIoData(t.Runtime).metatable)
	return c.PushingNext(t.Runtime, rt.UserDataValue(fv)), nil
}

func tostring(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	f, err := FileArg(c, 0)
	if err != nil {
		return nil, err
	}
	var s string
	if f.IsClosed() {
		s = "file (closed)"
	} else {
		s = fmt.Sprintf("file (%q)", f.Name())
	}
	t.RequireBytes(len(s))
	return c.PushingNext(t.Runtime, rt.StringValue(s)), nil
}

func newFileUserData(f *File, meta *rt.Table) *rt.UserData {
	return rt.NewUserData(f, meta)
}

File: lib/iolib/lua/iolib.lua
local function testf(f)
    return function(...)
        (function(ok, ...) 
            if ok then
                print("OK", ...)
            else
                print("ERR", ...)
            end
        end)(pcall(f, ...))
    end
end

do
    local close = testf(io.close)
    close("abc")
    --> ~ERR	.*must be a file

    print(pcall(io.open))
    --> ~^false\t.*value needed

    print(pcall(io.open, {}))
    --> ~^false\t.*must be a string

    print(pcall(io.open, "aaa", false))
    --> ~^false\t.*must be a string

    testf(io.open)("files/doesnotexist")
    --> ~OK\tnil\t

    local f = io.open("files/iotest.txt")
    print(f)
    --> =file ("files/iotest.txt")

    print(pcall(f.read))
    --> ~^false\t.*value needed

    print(pcall(f.read, 123))
    --> ~^false\t

    print(pcall(f.read, f, "?"))
    --> ~^false\t.*invalid format

    testf(io.type)()
    --> ~ERR	.*value needed

    print(io.type(123))
    --> =nil

    print(io.type(f))
    --> =file

    print(pcall(f.lines))
    --> ~^false\t.*value needed

    print(pcall(f.lines, 123))
    --> ~^false\t.*must be a file

    print(pcall(f.lines, f, "wat"))
    --> ~^false\t.*invalid format


    for line in f:lines() do
        print(line)
    end
    --> =hello
    --> =123
    --> =bye

    testf(f.close)()
    --> ~ERR	.*value needed

    testf(f.flush)()
    --> ~ERR	.*value needed

    testf(f.flush)(123)
    --> ~ERR	.*must be a file

    f:close()
    print(io.type(f))
    --> =closed file
end

do
    print(pcall(io.popen))
    --> ~^false\t.*value needed

    print(pcall(io.popen, {}))
    --> ~^false\t.*must be a string

    print(pcall(io.popen, "aaa", false))
    --> ~^false\t.*must be a string

    local f = io.popen("cat files/iotest.txt")
    print(f)
    --> =file ("cat files/iotext.txt")

    print(pcall(f.read))
    --> ~^false\t.*value needed

    print(pcall(f.read, 123))
    --> ~^false\t

    print(pcall(f.read, f, "?"))
    --> ~^false\t.*invalid format

    testf(io.type)()
    --> ~ERR	.*value needed

    print(io.type(123))
    --> =nil

    print(io.type(f))
    --> =file

    print(pcall(f.lines))
    --> ~^false\t.*value needed

    print(pcall(f.lines, 123))
    --> ~^false\t.*must be a file

    print(pcall(f.lines, f, "wat"))
    --> ~^false\t.*invalid format


    for line in f:lines() do
        print(line)
    end
    --> =hello
    --> =123
    --> =bye

    testf(f.close)()
    --> ~ERR	.*value needed

    testf(f.flush)()
    --> ~ERR	.*value needed

    testf(f.flush)(123)
    --> ~ERR	.*must be a file

    f:close()
    print(io.type(f))
    --> =closed file
end

do
    testf(io.lines)(123)
    --> ~ERR	.*must be a string

    testf(io.lines)("nonexistent")
    --> ~ERR	.*

    testf(io.lines)("files/iotest.txt", "z")
    --> ~ERR	.*invalid format

    for line in io.lines("files/iotest.txt") do
        print(line)
    end
    --> =hello
    --> =123
    --> =bye

    print((pcall(io.lines, "files/missing")))
    --> =false
end

do
    local function wp(x)
        print("[" .. tostring(x) .. "]")
    end
    local f = io.open("files/writetest.txt", "w")

    testf(f.write)()
    --> ~ERR	.*value needed

    print(pcall(f.write, 123))
    --> ~^false\t.*must be a file

    print(pcall(f.write, f, {}))
    --> ~^false\t.*must be a string or a number

    f:write("foobar", 1234, "\nabc\n")
    f:close()
    f = io.open("files/writetest.txt", "r")

    wp(f:read("a"))
    --> =[foobar1234
    --> =abc
    --> =]

    wp(f:read("l"))
    --> =[nil]

    testf(f.seek)(f, "set", "hello")
    --> ~ERR	.*must be an integer

    f:seek("set", 0)
    wp(f:read(7))
    --> =[foobar1]

    f:seek("cur", 3)
    wp(f:read(10))
    --> =[
    --> =abc
    --> =]

    f:seek("end", -4)
    wp(f:read("L"))
    --> =[abc
    --> =]

    f:seek("end", -2)
    f:seek("cur", -2)
    wp(f:read("L"))
    --> =[abc
    --> =]
    
    print(pcall(f.seek))
    --> ~false\t.*value needed

    print((pcall(f.seek, "hello")))
    --> =false

    print(pcall(f.seek, f, 42))
    --> ~^false\t.*string

    print(f:seek("set", -100000))
    --> ~^nil\t

    local metaf = getmetatable(f)

    testf(metaf.__tostring)()
    --> ~ERR	.*value needed

    testf(metaf.__tostring)("not a file")
    --> ~ERR	.*must be a file
end

do
    local function wp(x)
        print("[" .. tostring(x) .. "]")
    end
    local f = io.popen("cat > files/popenwrite.txt", "w")

    testf(f.write)()
    --> ~ERR	.*value needed

    print(pcall(f.write, 123))
    --> ~^false\t.*must be a file

    print(pcall(f.write, f, {}))
    --> ~^false\t.*must be a string or a number

    f:write("foobar", 1234, "\nabc\n")
    f:close()
    f = io.open("files/popenwrite.txt", "r")

    wp(f:read("a"))
    --> =[foobar1234
    --> =abc
    --> =]

    wp(f:read("l"))
    --> =[nil]
end

do
    testf(io.read)("z")
    --> ~ERR	.*invalid format

    testf(io.read)(false)
    --> ~ERR	.*invalid format

    local f = io.open("files/writetest2.txt", "w+")
    io.output(f)
    io.write([[Dear sir,
Blah blah,

Yours sincerely.
]])
    io.flush()
    io.input(f)
    f:seek("set", 0)
    print(io.read())
    --> =Dear sir,

    for line in io.lines() do
        print(line)
    end
    --> =Blah blah,
    --> =
    --> =Yours sincerely.
end

do
    local f = io.tmpfile()
    print(io.type(f))
    --> =file

    print(pcall(f.seek, f, "wat"))
    --> ~false\t.*: #1 must be "cur", "set" or "end"

    -- TODO: do something with the file
end

do
    -- local stdin = io.input()
    -- print(io.type(stdin))
    -- --> =file

    io.input("files/iotest.txt")
    print(io.input())
    --> =file ("files/iotest.txt")

    print((pcall(io.input, "files/missing")))
    --> =false

    print((pcall(io.input, 123)))
    --> =false
end

do
    -- local stdout = io.output()
    -- print(io.type(stdout))
    -- --> =file

    testf(io.output)(false)
    --> ~^ERR

    io.output("files/outputtest.txt")
    io.write("hello")
    io.write("bye")
    io.close()
    print(pcall(io.close))
    --> ~false\t.*file already closed
    print(io.open("files/outputtest.txt"):read())
    --> =hellobye
end

do
    local f = io.open("files/iotest.txt")
    print(f:read(0))
    --> =
    print(f:read("L"))
    --> =hello
    --> =
    print(f:read("n"))
    --> =123
    print(f:read("n"))
    --> =nil
    print(f:read("a"))
    --> =bye
    --> =
    print(f:read(0))
    --> =nil
end

do
    print(io.stdout:close())
    --> ~nil\t.*cannot close standard file

    print(io.stdin:close())
    --> ~nil\t.*cannot close standard file

    print(io.stderr:close())
    --> ~nil\t.*cannot close standard file
end

local ff -- this is to remember the file f below refers to after f goes out of scope.
do
    local f <close> = io.open("files/writetest3.txt", "w")
    ff = f
    -- hard to test what it does, at least make sure the correct modes are accepted
    f:setvbuf("no")
    f:setvbuf("full")
    f:setvbuf("line")
    f:setvbuf("full", 0)
    f:setvbuf("full", 2000)
    f:setvbuf("line", 1000)

    local function perr(...)
        print(pcall(function(...) f:setvbuf(...) end, ...))
    end

    perr("line", -1)
    --> ~false\t.*invalid buffer size

    perr("full", -1000)
    --> ~false\t.*invalid buffer size

    perr("blah")
    --> ~false\t.*invalid buffer mode

    perr()
    --> ~false\t.*2 arguments needed

    perr(100)
    --> ~false\t.*#2 must be a string

    perr("full", {})
    --> ~false\t.*#3 must be an integer

    -- ff is still open at this point
    print(io.type(ff))
    --> =file
end

-- New in Lua 5.4:
-- the __close metamethod was called, so ff should be closed now.
print(io.type(ff))
--> =closed file

do
    local close = getmetatable(ff).__close

    print(pcall(close))
    --> ~false\t.*value needed

    print(pcall(close, {}))
    --> ~false\t.*#1 must be a file
end
